druid-shell/src/platform/windows/application.rs

```diff
@@ -17,52 +17,86 @@
 use winapi::um::shellscalingapi::PROCESS_SYSTEM_DPI_AWARE;
 use winapi::um::wingdi::CreateSolidBrush;
 use winapi::um::winuser::{
-    DispatchMessageW, GetAncestor, GetMessageW, LoadIconW, PostQuitMessage, RegisterClassW,
-    TranslateAcceleratorW, TranslateMessage, GA_ROOT, IDI_APPLICATION, MSG, WNDCLASSW,
+    DispatchMessageW, GetAncestor, GetMessageW, LoadIconW, PostQuitMessage, PostThreadMessageW,
+    RegisterClassW, TranslateAcceleratorW, TranslateMessage, GA_ROOT, IDI_APPLICATION, MSG,
+    WNDCLASSW,
 };
 
+use log::{error, warn};
+
 use crate::application::AppHandler;
 
 use super::accels;
 use super::clipboard::Clipboard;
-use super::util::{self, ToWide, CLASS_NAME, OPTIONAL_FUNCTIONS};
-use super::window::win_proc_dispatch;
+use super::error::Error;
+use super::util::{
+    self, claim_main_thread, main_thread_id, release_main_thread, ToWide, CLASS_NAME,
+    OPTIONAL_FUNCTIONS,
+};
+use super::window::{win_proc_dispatch, DS_REQUEST_QUIT};
 
-pub struct Application;
+pub struct Application {
+    handler: Option<Box<dyn AppHandler>>,
+}
 
 impl Application {
-    pub fn new(_handler: Option<Box<dyn AppHandler>>) -> Application {
+    pub fn new(handler: Option<Box<dyn AppHandler>>) -> Application {
         Application::init();
-        Application
+        Application { handler }
     }
 
     pub fn run(&mut self) {
+        claim_main_thread();
         unsafe {
             // Handle windows messages
             loop {
                 let mut msg = mem::MaybeUninit::uninit();
                 let res = GetMessageW(msg.as_mut_ptr(), ptr::null_mut(), 0, 0);
                 if res <= 0 {
-                    return;
+                    if res == -1 {
+                        error!(
+                            "GetMessageW failed: {}",
+                            Error::Hr(HRESULT_FROM_WIN32(GetLastError()))
+                        );
+                    }
+                    break;
                 }
                 let mut msg: MSG = msg.assume_init();
                 let accels = accels::find_accels(GetAncestor(msg.hwnd, GA_ROOT));
                 let translated = accels.map_or(false, |it| {
                     TranslateAcceleratorW(msg.hwnd, it.handle(), &mut msg) != 0
                 });
-
                 if !translated {
-                    TranslateMessage(&msg);
-                    DispatchMessageW(&msg);
+                    // We check for DS_REQUEST_QUIT here because thread messages
+                    // will not be forwarded to any window procedures by DispatchMessageW.
+                    if msg.message == DS_REQUEST_QUIT {
+                        if let Some(handler) = &mut self.handler {
+                            // We want to queue up the destruction of all open windows.
+                            // Failure to do so will lead to resource leaks
+                            // and an eventual error code exit for the process.
+                            handler.close_all_windows();
```

Comment:
I wonder if we can invert this?

My naive vision: if you want to quit a druid app in windows, the druid layer is responsible for ensuring that all windows get closed, and _then_ we call into druid-shell and ask it to quit.

The inverse here feels difficult. `quit` in shell should not be the equivalent of a user-initiated quit; it should be a final request to drud-shell to terminate the process. This still feels like druid-shell is implementing policy.

One reason some of this stuff might be a bit weird is that there is a lingering assumption in a lot of this code that there's only ever one window, and a lot of our logic is oriented around that assumption.

Reply:
I think the `quit` in shell should work without issues. Calling it should never lead to resource leaks or errors. If we move the window closing decision to `druid` then all other users of `druid-shell` will either have to read the documentation and be sure to do the same, or they will have a resource leak by just using the `druid-shell` `quit`.

Reply:
It depends what the tradeoffs are.

Basically: if the tradeoff is that by making druid-shell be more difficult to use correctly, we allow druid to have a clearer and more intuitive API, it might be worth it.

`druid-shell` is hairy. It is 100% possible to misuse. In many ways, it makes sense to think of it as a C API.

If the documentation of `quit` clearly states that it should only be called when all windows are closed, I think that's reasonable. A compromise might be to have some mechanism in `druid-shell` to count windows, and then make `quit` return `Result<_>`, with an error if windows remain open; and an even better compromise might be to have the `Application` type become fuller featured, and include handles to open windows etc. But having us call `quit` and then it calling us back and asking us to close our windows feels like it is equally compromised, and it isn't clear to me that it offers any significant benefit.

Reply:
I'll play around with this a bit and see if I can perform the cleanup in another way without calling back to `druid`. The solution in this PR right now was just the quickest way to do it, as druid was already doing that bookkeeping. However I agree it's a bit spaghetti.

Now more generally, the `AppHandler` is passed to `Application` and right now only used for mac root menus. What kind of features do you think would roughly fit `AppHandler`? It's only calling back to `druid` right, but if we don't want to do that, when is it correct to use it? For some sort of event passing that isn't associated with any window? Cases where `druid-shell` is informing and doesn't depend on anyone listening or behaving in any specific way?

Reply:
So the genesis here: up until recently, druid-shell was entirely window based. That is, all events were associated with a window, and were delivered to a `WinHandler`.

This means that if you didn't have any open windows, we had no way to deliver events. In many cases this isn't a huge problem; but it really is on mac, which expects the application to remain open even without any windows.

So `AppHandler` is an attempt to introduce some type 'above' `WinHandler` that receives events at the 'application scope'.

Currently, yes, this is really just doing menus on mac. I could definitely imagine it having other roles; in particular it might make sense that the `AppHandler` gets notified when windows are added or removed (there is currently an `AppDelegate` for this, which predates `AppHandler`, and could perhaps be removed) and it may also make sense for `AppHandler` to receive certain system level events where it doesn't make sense to deliver them to a particular window. I can't think of great examples right now, but one might be if your app is registered to be able to open some type of file, and the user double clicks one of those files; we might notify your `AppHandler` at launch? 

In any case this didn't get a ton of long-term planning; it felt like something we would eventually need if we wanted to support windowless applications, and it was something that I clearly needed for runebender on mac.